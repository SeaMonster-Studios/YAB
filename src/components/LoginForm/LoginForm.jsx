import React from 'react'
import axios from 'axios'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import PropTypes from 'prop-types'
//
import { setAuthCookie } from '../../utils/auth/auth'
import * as authActionCreators from '../../store/actions/auth'

export class LoginForm extends React.Component {
  static propTypes = {
    setAuthenticatedUser: PropTypes.func.isRequired,
  }
  state = {
    password: {
      value: '',
    },
    email: {
      value: '',
    },
    errorCode: 0,
  }
  constructor(props) {
    super(props)
    this.handlePasswordChange = this.handlePasswordChange.bind(this)
    this.handleEmailChange = this.handleEmailChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }
  handlePasswordChange(event) {
    const password = event.target.value
    this.setState(prevState => ({
      ...prevState,
      password: {
        value: password,
      },
    }))
  }
  handleEmailChange(event) {
    const email = event.target.value
    this.setState(prevState => ({
      ...prevState,
      email: {
        value: email,
      },
    }))
  }

  async handleSubmit(e) {
    e.preventDefault()

    try {
      const { data } = await axios.post('/auth/login', {
        email: this.state.email.value,
        password: this.state.password.value,
      })
      this.props.setAuthenticatedUser({
        authed: data.auth,
        user: {
          email: this.state.email.value,
          id: data.id,
        },
      })
      setAuthCookie(data.token)
    } catch (error) {
      if (error.response) {
        this.setState(prevState => ({
          ...prevState,
          errorCode: error.response.status,
        }))
      } else {
        // SENTRY will handle this
        console.error(error)
      }
    }
  }
  render() {
    return (
      <form onSubmit={this.handleSubmit} data-testid="login-form">
        <input
          type="email"
          placeholder="Email"
          name="email"
          data-testid="login-form-email"
          value={this.state.email.value}
          required
          onChange={this.handleEmailChange}
        />
        <input
          type="password"
          required
          name="password"
          value={this.state.password.value}
          data-testid="login-form-password"
          placeholder="Password"
          onChange={this.handlePasswordChange}
        />
        <button type="submit">Login</button>
        <FormError errorCode={this.state.errorCode} />
      </form>
    )
  }
}

function FormError({ errorCode }) {
  if (errorCode === 404) {
    return (
      <div className="formError" data-testid="form-error-404">
        Sorry, but there is no account registered with that email address.
      </div>
    )
  } else if (errorCode > 0) {
    return (
      <div className="formError" data-testid="form-error-401">
        Sorry, but the email/password combination you entered does not exist.
      </div>
    )
  }
  return null
}

FormError.propTypes = {
  errorCode: PropTypes.number.isRequired,
}

const mapStateToProps = () => ({})

const mapDispatchToProps = dispatch =>
  bindActionCreators(authActionCreators, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(LoginForm)
