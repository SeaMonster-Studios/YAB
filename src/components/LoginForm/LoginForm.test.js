import React from 'react'
import { Simulate, wait, render } from 'react-testing-library'
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import faker from 'faker'
//
import { renderWithRedux } from '../../utils/testing/testing'
import LoginForm, { LoginForm as LoginFormSansRedux } from './LoginForm'

describe('LoginForm Component Test', () => {
  it('Renders with correct default values from redux store', () => {
    const wrapper = renderSetupWithRedux()

    // make sure it renders
    expect(wrapper.container).toMatchSnapshot()

    // No user should be logged in if this form is showing
    expect(wrapper.initialState.auth.authed).toBe(false)
    expect(wrapper.initialState.auth.user.email).toBe('')
    expect(wrapper.initialState.auth.user.id).toBe('')
  })

  it('Expects email input value to update on user input', () => {
    let { email } = renderSetup()
    const value = faker.random.words()
    Simulate.change(email, { target: { value } })

    expect(email.value).toBe(value)
  })

  it('Expects password input value to update on user input', () => {
    let { password } = renderSetup()
    const value = faker.random.words()
    Simulate.change(password, { target: { value } })

    expect(password.value).toBe(value)
  })

  it('Expects a relevant error message to be displayed when no user is found', async () => {
    new MockAdapter(axios).onPost('/auth/login').reply(404)

    const wrapper = renderSetupWithRedux()
    const { email, password, getByTestId } = wrapper
    let formError

    Simulate.change(password, { target: { value: 'password' } })
    Simulate.change(email, {
      target: { value: 'someonefkjaskdfjfj@gmail.com' },
    })

    await wait(() => Simulate.submit(getByTestId('login-form')))
    const stateAfterSubmit = wrapper.store.getState()

    await wait(() => (formError = wrapper.getByTestId('form-error-404')))

    expect(formError).toBeDefined()
    expect(wrapper.container).toMatchSnapshot()
    expect(stateAfterSubmit.auth.authed).toBe(false)
  })

  it('Expects a relevant error message to be displayed when password is incorrect', async () => {
    new MockAdapter(axios).onPost('/auth/login').reply(401)

    const wrapper = renderSetupWithRedux()
    const { email, password, getByTestId } = wrapper
    let formError

    Simulate.change(password, { target: { value: 'password' } })
    Simulate.change(email, {
      target: { value: 'someonefkjaskdfjfj@gmail.com' },
    })

    await wait(() => Simulate.submit(getByTestId('login-form')))
    const stateAfterSubmit = wrapper.store.getState()

    await wait(() => (formError = wrapper.getByTestId('form-error-401')))

    expect(formError).toBeDefined()
    expect(wrapper.container).toMatchSnapshot()
    expect(stateAfterSubmit.auth.authed).toBe(false)
  })

  it('Expects a default error message to be displayed when there is a 500 error', async () => {
    new MockAdapter(axios).onPost('/auth/login').reply(500)

    const wrapper = renderSetupWithRedux()
    const { email, password, getByTestId } = wrapper
    let formError

    Simulate.change(password, { target: { value: 'password' } })
    Simulate.change(email, {
      target: { value: 'someonefkjaskdfjfj@gmail.com' },
    })

    await wait(() => Simulate.submit(getByTestId('login-form')))
    const stateAfterSubmit = wrapper.store.getState()

    await wait(() => (formError = wrapper.getByTestId('form-error-401')))

    expect(formError).toBeDefined()
    expect(wrapper.container).toMatchSnapshot()
    expect(stateAfterSubmit.auth.authed).toBe(false)
  })
})

function renderSetupWithRedux(overrides) {
  const wrapper = renderWithRedux(<LoginForm />)
  const email = wrapper.getByTestId('login-form-email')
  const password = wrapper.getByTestId('login-form-password')
  const initialState = wrapper.store.getState()

  return {
    ...wrapper,
    email,
    password,
    initialState,
    ...overrides,
  }
}

function renderSetup(overrides) {
  const wrapper = render(
    <LoginFormSansRedux setAuthenticatedUser={() => ({})} />,
  )
  const email = wrapper.getByTestId('login-form-email')
  const password = wrapper.getByTestId('login-form-password')

  return {
    ...wrapper,
    email,
    password,
    ...overrides,
  }
}
