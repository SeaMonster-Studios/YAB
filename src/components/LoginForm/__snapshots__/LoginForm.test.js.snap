// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Component Test Expects a default error message to be displayed when there is a 500 error 1`] = `
<div>
  <form
    data-testid="login-form"
  >
    <input
      data-testid="login-form-email"
      name="email"
      placeholder="Email"
      required=""
      type="email"
      value="someonefkjaskdfjfj@gmail.com"
    />
    <input
      data-testid="login-form-password"
      name="password"
      placeholder="Password"
      required=""
      type="password"
      value="password"
    />
    <button
      type="submit"
    >
      Login
    </button>
    <div
      class="formError"
      data-testid="form-error-401"
    >
      Sorry, but the email/password combination you entered does not exist.
    </div>
  </form>
</div>
`;

exports[`LoginForm Component Test Expects a relevant error message to be displayed when no user is found 1`] = `
<div>
  <form
    data-testid="login-form"
  >
    <input
      data-testid="login-form-email"
      name="email"
      placeholder="Email"
      required=""
      type="email"
      value="someonefkjaskdfjfj@gmail.com"
    />
    <input
      data-testid="login-form-password"
      name="password"
      placeholder="Password"
      required=""
      type="password"
      value="password"
    />
    <button
      type="submit"
    >
      Login
    </button>
    <div
      class="formError"
      data-testid="form-error-404"
    >
      Sorry, but there is no account registered with that email address.
    </div>
  </form>
</div>
`;

exports[`LoginForm Component Test Expects a relevant error message to be displayed when password is incorrect 1`] = `
<div>
  <form
    data-testid="login-form"
  >
    <input
      data-testid="login-form-email"
      name="email"
      placeholder="Email"
      required=""
      type="email"
      value="someonefkjaskdfjfj@gmail.com"
    />
    <input
      data-testid="login-form-password"
      name="password"
      placeholder="Password"
      required=""
      type="password"
      value="password"
    />
    <button
      type="submit"
    >
      Login
    </button>
    <div
      class="formError"
      data-testid="form-error-401"
    >
      Sorry, but the email/password combination you entered does not exist.
    </div>
  </form>
</div>
`;

exports[`LoginForm Component Test Renders with correct default values from redux store 1`] = `
<div>
  <form
    data-testid="login-form"
  >
    <input
      data-testid="login-form-email"
      name="email"
      placeholder="Email"
      required=""
      type="email"
      value=""
    />
    <input
      data-testid="login-form-password"
      name="password"
      placeholder="Password"
      required=""
      type="password"
      value=""
    />
    <button
      type="submit"
    >
      Login
    </button>
  </form>
</div>
`;
